packages/query/src/block-processor.ts

```diff
@@ -0,0 +1,367 @@
+    // - write to idb.
+    for (const nullifier of nullifiers) {
+      const record =
+        (await this.indexedDb.getNoteByNullifier(nullifier)) ??
+        (await this.indexedDb.getSwapByNullifier(nullifier));
+
+      if (!record) continue;
+
+      // resolve nullifiers for any records of any source type
+      if ('heightSpent' in record) {
+        // spendable note
+        record.heightSpent = height;
+        await this.indexedDb.saveSpendableNote(record);
+      } else if ('heightClaimed' in record) {
+        // swap claimed
+        record.heightClaimed = height;
+        await this.indexedDb.saveSwap(record);
+      }
+
+      // we want to recover ids for transaction sources that don't have an id,
+      // represented by an empty Uint8Array
+      if (
+        record.source?.equals({ source: { case: 'transaction', value: { id: new Uint8Array() } } })
+      )
+        recordNeedsSourceTxId.set(uint8ArrayToHex(nullifier.inner), record);
+    }
+
+    return recordNeedsSourceTxId;
+  }
+
+  // TODO: can we use compactBlock.nullifiers?
+  async resolveSources(
+    sourcelessRecords: Map<string, SwapRecord | SpendableNoteRecord>,
+    blockTransactions: Transaction[],
+  ) {
+    // map all of this block's tx action nullifiers to ids of the containing tx.
+    // nullifier inner as matchable string -> transaction
+    const blockTxByNullifier = new Map<string, Transaction>();
+
+    for (const tx of blockTransactions) {
+      const actionsNullifiers = tx.body?.actions.map(({ action }) => {
+        switch (action.case) {
+          case 'spend':
+          case 'swapClaim':
+            return action.value.body?.nullifier;
+          default:
+            return;
+        }
+      });
+      actionsNullifiers?.map(
+        nullifier =>
+          nullifier?.inner.length && blockTxByNullifier.set(uint8ArrayToHex(nullifier.inner), tx),
+      );
+    }
+
+    // match record nullifiers to those tx nullifiers
+    // if we have a tx, we can hash it for id, and assign source
+    // collect txs we use this way, to generate transactioninfos later
+    const relevantTx = new Map<string, Transaction>();
+    for (const [nullifierString, record] of sourcelessRecords.entries()) {
+      const sourceTx = blockTxByNullifier.get(nullifierString);
+      if (!sourceTx) {
+        console.error("no tx on block for nullifier on block, can't generate transactioninfo");
+        continue;
+      }
+
+      const id = await sha256Hash(sourceTx.toBinary());
+      // TODO: is there a race condition here? should we query the record again?
+      record.source = new CommitmentSource({
+        source: { case: 'transaction', value: { id } },
+      });
+      const idString = uint8ArrayToHex(id);
+      relevantTx.set(idString, sourceTx);
+
+      if (record instanceof SpendableNoteRecord) await this.indexedDb.saveSpendableNote(record);
+      else if (record instanceof SwapRecord) await this.indexedDb.saveSwap(record);
+    }
+
```

Comment:
We missed the definition of transactions relevant by commitment. 
We should compare the commitment from `new_notes` and `new_swaps` with the commitment in `tx.actions` 
https://github.com/penumbra-zone/penumbra/blob/2b0408879aaef969aade53cb93cff1dad69b99ce/crates/view/src/worker.rs#L155C42-L155C42 
In this case we must define source for `new_notes` and `new_swaps`